### PR TITLE
chore: creating meta package for default auto instrumentations for node

### LINKED ISSUE #377: Create meta package for auto instrumentations for node

### PR DESCRIPTION
**Pull request by obecny**
## Which problem is this PR solving?

- Fixes #377 

## Short description of the changes

- Creates meta package for default auto instrumentations for node

Usage
```
const { NodeTracerProvider } = require('@opentelemetry/node');
const { getNodeAutoInstrumentations } = require('@opentelemetry/auto-instrumentations-node');
const { CollectorTraceExporter } = require('@opentelemetry/exporter-collector');
const { SimpleSpanProcessor } = require('@opentelemetry/tracing');
const { registerInstrumentations } = require('@opentelemetry/instrumentation');

const exporter = new CollectorTraceExporter({
  serviceName: 'auto-instrumentations-node',
});

const provider = new NodeTracerProvider();
provider.addSpanProcessor(new SimpleSpanProcessor(exporter));
provider.register();

registerInstrumentations({
  instrumentations: [
    getNodeAutoInstrumentations({
      // load custom configuration for http instrumentation
      "@opentelemetry/instrumentation-http": {
        applyCustomAttributesOnSpan: (span)=> {
          span.setAttribute('foo2', 'bar2');
        },
      },
    }),
  ],
});
```

### CODE REVIEW

**Comment by obecny** [comment]
@open-telemetry/javascript-approvers this is basically complete but without tests, please have a look before I start adding unit tests

File: examples/meta-node/tracer.js
Comment (vmarchaud) [inline examples/meta-node/tracer.js:28]: do you get the typing of available options here ? I think it would be important to have them

Comment (obecny) [reply examples/meta-node/tracer.js:28]: no, neither this is possible now with old plugin loader

Comment (vmarchaud) [reply examples/meta-node/tracer.js:28]: I think it could be much more easier for users if they could be typesafe, its merely just a map that gives out the type of the config for a given instrumentation name like so:
```ts
type InstrumentationConfigMap = {
  '@opentelemetry/instrumentation-http': HttpInstrumentationConfig
  etc ...
}
```

Comment (obecny) [reply examples/meta-node/tracer.js:28]: fixed

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (dyladan) [inline metapackages/auto-instrumentations-node/src/utils.ts:26]: Why is this one different than all the others

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (dyladan) [inline metapackages/auto-instrumentations-node/src/utils.ts:105]: Maybe easier to manage a single object which has the instrumentation and the config?

```typescript
const InstrumentationMap = {
    '@opentelemetry/instrumentation-dns': {
        instrumentation: DnsInstrumentation,
        config: { enabled: true } as DnsInstrumentationConfig,
    },
    '@opentelemetry/instrumentation-express': {
        instrumentation: ExpressInstrumentation,
        config: { enabled: true } as ExpressInstrumentationConfig,
    },
    // etc
}

type InstrumentationConfigMap = {
    [Name in keyof typeof InstrumentationMap]?: typeof InstrumentationMap[Name]["config"]
};
```

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (dyladan) [inline metapackages/auto-instrumentations-node/src/utils.ts:62]: ~~This can be done using `keyof InstrumentationMap` so you don't have to update in so many places~~

edit: I actually think all of these types can be compressed into a single place to manage

**Review by dyladan — COMMENTED** [review]
This looks great, just have a suggestion for a way to lower the number of places that need to be updated when we add a package.

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (dyladan) [inline metapackages/auto-instrumentations-node/src/utils.ts:122]: If you implement my suggestion, this will need to be built by looping over `Object.keys(InstrumentationMap)`

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (dyladan) [reply metapackages/auto-instrumentations-node/src/utils.ts:105]: If you would like I can make a PR into your branch with these changes applied

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (dyladan) [reply metapackages/auto-instrumentations-node/src/utils.ts:26]: Can we fix this so they are all the same?

Comment (obecny) [reply metapackages/auto-instrumentations-node/src/utils.ts]: leaving for now the basic config, can be changed after releasing the core as grpc is part of core :/

File: metapackages/auto-instrumentations-node/src/utils.ts
Comment (obecny) [reply metapackages/auto-instrumentations-node/src/utils.ts:108]: Tried that but it was still complaining with some type when trying to create a defaultConfig - it was more and more casting so I finally decided to not complicate it anymore to be easy to change it in future. I think it can be still refactored later if you want, WDYT ?

Comment (dyladan) [reply metapackages/auto-instrumentations-node/src/utils.ts:108]: Here is what i'm talking about. If you don't like it feel free to close it https://github.com/obecny/opentelemetry-js-contrib/pull/2

File: metapackages/auto-instrumentations-node/README.md
Comment (dyladan) [inline metapackages/auto-instrumentations-node/README.md:65]:
```suggestion
[npm-url]: https://www.npmjs.com/package/@opentelemetry/auto-instrumentations-node
[npm-img]: https://badge.fury.io/js/%40opentelemetry%2Fauto-instrumentations-node.svg
```

File: metapackages/auto-instrumentations-node/README.md
Comment (dyladan) [inline metapackages/auto-instrumentations-node/README.md:63]:
```suggestion
[dependencies-image]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib.svg?path=packages%2Fauto-instrumentations-node
[dependencies-url]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib?path=packages%2Fauto-instrumentations-node
[devDependencies-image]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib.svg?path=packages%2Fauto-instrumentations-node&type=dev
[devDependencies-url]: https://david-dm.org/open-telemetry/opentelemetry-js-contrib?path=packages%2Fauto-instrumentations-node&type=dev
```

File: metapackages/auto-instrumentations-node/README.md
Comment (dyladan) [inline metapackages/auto-instrumentations-node/README.md:56]: we're no longer using gitter

File: metapackages/auto-instrumentations-node/package.json
Comment (dyladan) [inline metapackages/auto-instrumentations-node/package.json:6]: homepage is core not contrib. is the intended?

File: metapackages/auto-instrumentations-node/package.json
Comment (dyladan) [inline metapackages/auto-instrumentations-node/package.json:46]: 0.18.1 is latest, but rc.0 will be latest pretty soon. We should probably hold off rc.0 until we can update the whole repo at once

```suggestion
    "@opentelemetry/api": "^0.18.1",
```

Comment (obecny) [reply metapackages/auto-instrumentations-node/package.json:46]: this should be updated once for all repo before release, there are already other PRS that are merged pointing to `0.18.0`

**Review by dyladan — APPROVED** [review]

**Review by vmarchaud — APPROVED** [review]

**Comment by vmarchaud** [comment]
@obecny Do you want to wait for mysql plugin to release this ?

**Comment by obecny** [comment]
> @obecny Do you want to wait for mysql plugin to release this ?

it all depends when the mysql will be merged / done

**Pull request merged**
